service/topology_coordinator.cc

```diff
@@ -546,41 +546,48 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
     future<> clean_obsolete_cdc_generations(
             const group0_guard& guard,
             std::vector<canonical_mutation>& updates,
             sstring& reason) {
-        auto candidate = co_await _sys_ks.get_cdc_generations_cleanup_candidate();
-        if (!candidate) {
+        const auto& committed_gens = _topo_sm._topology.committed_cdc_generations;
+        if (committed_gens.empty()) {
             co_return;
         }
 
-        // We cannot delete the current CDC generation. We must also ensure that timestamps of all deleted
-        // generations are in the past compared to all nodes' clocks. Checking that the clean-up candidate's
-        // timestamp does not exceed now() - 24 h should suffice with a safe reserve. We don't have to check
-        // the timestamps of other CDC generations we are removing because the candidate's is the latest
-        // among them.
+        // If some node can still accept a write to a CDC generation, we cannot delete this generation.
+        // A request coordinator accepts a write to one of the previous generations (the ones that stopped
+        // operating before `now`) if the write's timestamp is higher than `now - 5s` where `now` is
+        // provided by the coordinator's local clock. So, we can safely delete a generation if it
+        // stopped operating more than 5 s ago on all nodes. Since their clocks can be desynchronized,
+        // we delete only generations that stopped operating more than 24 hours ago, which is a safe and
+        // sufficient choice.
         auto ts_upper_bound = db_clock::now() - std::chrono::days(1);
-        utils::get_local_injector().inject("clean_obsolete_cdc_generations_ignore_ts", [&] {
-            ts_upper_bound = candidate->ts;
+        utils::get_local_injector().inject("clean_obsolete_cdc_generations_change_ts_ub", [&] {
+            ts_upper_bound = db_clock::now();
         });
-        if (candidate == _topo_sm._topology.committed_cdc_generations.back() || candidate->ts > ts_upper_bound) {
+
+        std::optional<std::vector<cdc::generation_id_v2>::const_iterator> first_nonobsolete_gen_it;
+        for (auto it = committed_gens.begin(); it != committed_gens.end() && it->ts <= ts_upper_bound; it++) {
+            if (it + 1 == committed_gens.end() || (it + 1)->ts > ts_upper_bound) {
+                first_nonobsolete_gen_it = it;
+            }
+        }
+        if (!first_nonobsolete_gen_it || *first_nonobsolete_gen_it == committed_gens.begin()) {
             co_return;
         }
 
         auto mut_ts = guard.write_timestamp();
 
-        // Mark the lack of a new clean-up candidate. The current one will be deleted.
-        mutation m = _sys_ks.make_cleanup_candidate_mutation(std::nullopt, mut_ts);
-
-        // Insert a tombstone covering all generations that have time UUID not higher than the candidate.
+        // Insert a tombstone covering all obsolete generations.
```

Comment:
Hmm, I think it's theoretically possible that we may clean data of generation that was not published yet, which would be a big problem. At least I see no mechanism to protect us from this. We should probably stop our first_nonobsolete_gen_it search when we see that the gen is unpublished yet.

But IIUC this is a preexisting problem, it was present with the cleanup_candidate as well. And the current PR only improves the situation because it clears generations later. So we can fix in a follow-up, it should be a simple fix

Reply:
Yes, this is a preexisting problem. I'll send a follow-up.